### PR TITLE
Fix cpp api python filter

### PR DESCRIPTION
**Pull request by dbenage-cx**
Workaround for the multiple no file found warnings:
 `sh: @PYTHON_EXECUTABLE@ @FreeOrion_SOURCE_DIR@/doc/tag_filter.py: No such file or directory
`

Related #1254

### CODE REVIEW

**Comment by Cjkjvfnby** [comment]
Does docs work on windows?

**Comment by dbenage-cx** [comment]
@Cjkjvfnby not to my knowledge unless your using cmake, which I assume would be `Cygwin` or `MinGW`

**Comment by dbenage-cx** [comment]
As it was no longer working, rewrote the tag filter allowing for full comment blocks in python.

File: doc/tag_filter.py [hunk 1]
```diff
@@ -4,26 +4,36 @@
...
-while line and tag_match:
-    # doxygen wont document this alias if it is indented
-    if tag_comment in line:
-        line = line.strip() + '\n'
-    sys.stdout.write(line)
-    line = file_obj.readline()
-file_obj.close()
+# Output the entire file if it contains a match
+# entire file is needed for context
+if match_found:
```
Comment (Cjkjvfnby) [inline doc/tag_filter.py:37]: Drop `match_found` and this check.  IF buffer is empty nothing will be written to stdout.

I prefer to see something like 
```
sys.stdout.write('\n'.join(buffer))
sys.stdout.write('\n')
```

File: doc/tag_filter.py [hunk 2]
```diff
@@ -4,26 +4,36 @@
...
+# dedent doxygen tag documentation
+with open(file_name, 'r') as f:
+    # sys.stderr.write("checking file: " + file_name + '\n')
+    for line in f:
+        # doxygen wont document this alias if it is indented
+        if tag_comment in line:
+            ln_void, ln_tag, ln_content = line.partition(tag_comment)
+            prev_line.append('## ' + tag_comment + ln_content.rstrip('\n'))
+            match_found = True
+        elif prev_line:
```
Comment (Cjkjvfnby) [inline doc/tag_filter.py]: Maybe rename to something like `collected_data`

File: doc/tag_filter.in [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    exit
+fi
+
+py_exec="@PYTHON_EXECUTABLE@"
+py_sh="@FreeOrion_SOURCE_DIR@/doc/tag_filter.py"
+
+
+if [[ -r "${target_file}" && -x "${py_exec}" && -r "${py_sh}" ]];
+then
+    eval "$py_exec $py_sh $target_file"
```
Comment (Cjkjvfnby) [inline doc/tag_filter.in]: And who is catching stdout?

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
I checked python code.
I cant run this code. 

It there a place where I can check documentation?

Here is my attempt for python documentation
https://github.com/Cjkjvfnby/freeorion/pull/15

File: doc/tag_filter.in (hunk 3 above)
Comment (dbenage-cx) [reply doc/tag_filter.in]: Doxygen,  I will explain more in a comment outside of this review block (so others may see it easier)

File: doc/tag_filter.py (hunk 1 above)
Comment (dbenage-cx) [reply doc/tag_filter.py:37]: The purpose of `match_found` is to not output the buffer when there is no match.
Unless we want to document all python code within the cpp api?

File: doc/tag_filter.py (hunk 2 above)
Comment (dbenage-cx) [reply doc/tag_filter.py]: Fixed

File: doc/tag_filter.py [hunk 4]
```diff
@@ -4,26 +4,36 @@
...
 
-tag_comment = "content_tag{"
-tag_match = 0
+# dedent doxygen tag documentation
+with open(file_name, 'r') as f:
+    # sys.stderr.write("checking file: " + file_name + '\n')
+    for line in f:
+        # doxygen wont document this alias if it is indented
+        if tag_comment in line:
+            ln_void, ln_tag, ln_content = line.partition(tag_comment)
```
[the remark this answers is not in the document]
Comment (dbenage-cx) [reply doc/tag_filter.py]: Fixed

File: doc/tag_filter.py [hunk 5]
```diff
@@ -4,26 +4,36 @@
...
+prev_line = list()
+buff = list()
 
-tag_comment = "content_tag{"
-tag_match = 0
+# dedent doxygen tag documentation
+with open(file_name, 'r') as f:
+    # sys.stderr.write("checking file: " + file_name + '\n')
+    for line in f:
+        # doxygen wont document this alias if it is indented
```
[the remark this answers is not in the document]
Comment (dbenage-cx) [reply doc/tag_filter.py]: Fixed

File: doc/tag_filter.py [hunk 6]
```diff
@@ -4,26 +4,36 @@
 import os
 
 file_name = sys.argv[1]
-file_obj = open(file_name, 'r')
+tag_comment = "@content_tag{"
+match_found = False
+prev_line = list()
```
[the remark this answers is not in the document]
Comment (dbenage-cx) [reply doc/tag_filter.py]: Fixed

**Comment by dbenage-cx** [comment]
@Cjkjvfnby TY for reviewing python portion.

~~This helper script is translated from `doc/tag_filter.in` by CMake.~~
**Edit**: Removed helper script, config line was not properly escaped.

Doxygen is configured to call this script for any `*.py` file, with the input file as the argument
 ( `tag_filter.sh` `source_file.py` -> `python` `tag_filter.py` `source_file.py` ).
The output is then used to determine content and link creation within the generated documentation.
If there is any output, lines can not be added or removed, otherwise anchors are not correctly placed.
When there is no output, there is no generated content.

The purpose of including some of the python code here is to provide a central place for documentation of all `content definition tags`.  The specific page for these tags is: [Content Definition Tags](http://freeorion.github.io/cppapi/master/content_tags.html).  FOCS files are not yet parsed.

> It there a place where I can check documentation?

I can upload a generated cpp-api from this branch if needed, but the main concerns addressed here are not reflected in the current output of master:
* This PR allows cases such as [this change](https://github.com/freeorion/freeorion/pull/1225#pullrequestreview-17138212) to still be valid.
* More efficient and correct python implementation
* Reduces error warnings when generating the documentation

> Here is my attempt for python documentation

For the purpose of content tag documentation, I do not have a preference between Doxygen and Sphinx.  For a central reference of these tags, c++, python, and definition script files will need to be parsed.  (I will put Sphinx on my todo list to familiarize myself with)

> Does docs work on windows?

They could be made to work with an installed windows Doxygen binary and manually changing the variables to proper locations in `doc/Doxyfile.in` (or providing some method to do so).

**Comment by adrianbroher** [comment]
> Does docs work on windows?

Generate your platform specific build system with CMake and run the `doc` target.  This will create the documentation for C++ with Doxygen.  Opposing to what @dbenage-cx said this isn't limited (or related at all) to `Cygwin` or `MinGW`.  This is not related to your attempt of documentation generation for python.

> It there a place where I can check documentation?

http://freeorion.github.io/cppapi/master/

> Here is my attempt for python documentation.

I am aware of your efforts, but realizing this as part of the CI isn't as easy as you may think.  As far as I know currently there is no way to generate the python documentation without running the freeorion client without human interaction.

> I cant run this code.

As @dbenage-cx already mentioned it is not intended to be run as standalone program, but rather as a [doxygen input filter](http://www.stack.nl/~dimitri/doxygen/manual/config.html#cfg_input_filter).

**Comment by adrianbroher** [comment]
@dbenage-cx 

> fixup 3ef3906

Is there are reason why this isn't fixed up in git (e.g. squashed)?

**Comment by dbenage-cx** [comment]
@adrianbroher : reviewer preference, I plan to fixup that commit once changes are approved.

File: doc/tag_filter.py (hunk 1 above)
Comment (Cjkjvfnby) [reply doc/tag_filter.py:37]: Now I see, that you rebuild full file. I thought that you extract only comments to console.

**Review by Cjkjvfnby — APPROVED** [review]

**Comment by adrianbroher** [comment]
Works for me.  The reported error is gone.

**Pull request merged**